cwl/src/main/scala/wdl4s/cwl/CwlFile.scala

```diff
@@ -0,0 +1,262 @@
+                            requirements: Option[Array[Requirement]] = None,
+                            //hints: Option[Array[CwlAny]] = None,
+                            hints: Option[Array[Map[String, String]]] = None,
+                            label: Option[String] = None,
+                            doc: Option[String] = None,
+                            cwlVersion: Option[CwlVersion] = Option(CwlVersion.Version1),
+                            baseCommand: Option[BaseCommand] = None,
+                            arguments: Option[Array[CommandLineTool.Argument]] = None,
+                            stdin: Option[StringOrExpression] = None,
+                            stderr: Option[StringOrExpression] = None,
+                            stdout: Option[StringOrExpression] = None,
+                            successCodes: Option[Array[Int]] = None,
+                            temporaryFailCodes: Option[Array[Int]] = None,
+                            permanentFailCodes: Option[Array[Int]] = None) extends CwlFile {
+
+  def womExecutable: ErrorOr[Executable] =
+    Valid(Executable(taskDefinition))
+
+
+  object BaseCommandToString extends Poly1 {
+    implicit def one = at[String] {
+      identity
+    }
+
+    implicit def many = at[Array[String]] {
+      _.mkString(" && ")
+    }
+  }
+
+  object ArgumentToId extends Poly1 {
+    implicit def ecmaScript = at[ECMAScriptExpression] {
+      _.value
+    }
+
+    implicit def commandLineBinding = at[CommandLineBinding] { _ => "" }
+
+    implicit def string = at[String] {
+      identity
+    }
+  }
+
+  /**
+    * This is used in place of the id when id is None.
+    *
+    * @return
+    */
+  def taskDefinitionId: String =
+    baseCommand.map(_.fold(BaseCommandToString)).getOrElse(
+      arguments.map(_.map(_.fold(ArgumentToId)).mkString(" ")).get)
+
+  def taskDefinition: TaskDefinition = {
+
+    val id = this.id.getOrElse(taskDefinitionId)
+
+    val commandTemplate: Seq[CommandPart] = baseCommand.get.fold(BaseCommandToCommandParts)
+
+    val runtimeAttributes: RuntimeAttributes = RuntimeAttributes(Map.empty[String, WdlExpression])
+
+    val meta: Map[String, String] = Map.empty
+    val parameterMeta: Map[String, String] = Map.empty
+
+    //TODO: This output does _not_ capture expressions from the output.outputBinding
+    //The implementation must include the expression evaluation pieces as detailed in:
+    //http://www.commonwl.org/v1.0/CommandLineTool.html#CommandOutputBinding
+    val outputs: Set[Callable.OutputDefinition] = this.outputs.map {
+      output =>
+        val tpe = output.`type`.flatMap(_.select[CwlType]).map(cwlTypeToWdlType).get //<-- here be `get` dragons
+        OutputDefinition(output.id, tpe, PlaceholderWomExpression(Set.empty, tpe))
+    }.toSet
+
+    val inputs: Set[_ <: Callable.InputDefinition] =
+      this.inputs.map { cip =>
+        val tpe = cip.`type`.flatMap(_.select[CwlType]).map(cwlTypeToWdlType).get
```

Comment:
`wdlType` please :) Or a type ascription if you prefere but so it's clear what type of type it is just by glancing at it